## Import required libraries

In [1]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
import xgboost as xgb
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, f1_score
import pickle

## Load the data

In [2]:
X_train_scaled = pd.read_csv('../data/final/X_train_scaled.csv')
X_test_scaled = pd.read_csv('../data/final/X_test_scaled.csv')
y_train = pd.read_csv('../data/final/y_train.csv')
y_test = pd.read_csv('../data/final/y_test.csv')

## Logistic Regression

In [3]:
model = LogisticRegression()
model.fit(X_train_scaled, y_train)

y_pred = model.predict(X_test_scaled)

/opt/anaconda3/envs/llm_env/lib/python3.10/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [4]:
#with open('../data/models/feature_importance_LogReg(0.74).pkl', 'wb') as file:
#    pickle.dump(model, file)

In [5]:
# Accuracy & F1 Score
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')
print(f'f1_score: {f1:.2f}')

Accuracy: 0.65
f1_score: 0.73


In [6]:
# Classification Report
class_report = classification_report(y_test, y_pred, target_names=['Down', 'Up'])
print('Classification Report:')
print(class_report)

Classification Report:
              precision    recall  f1-score   support

        Down       0.83      0.36      0.50       321
          Up       0.60      0.93      0.73       341

    accuracy                           0.65       662
   macro avg       0.72      0.64      0.61       662
weighted avg       0.71      0.65      0.62       662



## Random Forrest Classifier

In [7]:
# Train the random forest classifier
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train_scaled, y_train)

# Make predictions
y_pred = model.predict(X_test_scaled)

/opt/anaconda3/envs/llm_env/lib/python3.10/site-packages/sklearn/base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


In [8]:
#with open('../data/models/version7_randomforest(0.61).pkl', 'wb') as file:
#    pickle.dump(model, file)

In [9]:
# Accuracy & F1 Score
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')
print(f'f1_score: {f1:.2f}')

Accuracy: 0.66
f1_score: 0.67


In [10]:
# Classification Report
class_report = classification_report(y_test, y_pred, target_names=['Down', 'Up'])
print('Classification Report:')
print(class_report)

Classification Report:
              precision    recall  f1-score   support

        Down       0.65      0.67      0.66       321
          Up       0.68      0.65      0.67       341

    accuracy                           0.66       662
   macro avg       0.66      0.66      0.66       662
weighted avg       0.66      0.66      0.66       662



## Support Vector Machines

In [11]:
model = SVC()
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)

/opt/anaconda3/envs/llm_env/lib/python3.10/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [12]:
#with open('../data/models/version11.1_SVM(0.69).pkl', 'wb') as file:
#    pickle.dump(model, file)

In [13]:
# Accuracy & F1 Score
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')
print(f'f1_score: {f1:.2f}')

Accuracy: 0.55
f1_score: 0.47


In [14]:
# Classification Report
class_report = classification_report(y_test, y_pred, target_names=['Down', 'Up'])
print('Classification Report:')
print(class_report)

Classification Report:
              precision    recall  f1-score   support

        Down       0.52      0.71      0.60       321
          Up       0.59      0.40      0.47       341

    accuracy                           0.55       662
   macro avg       0.56      0.55      0.54       662
weighted avg       0.56      0.55      0.54       662



## k-Nearest Neighbors (k-NN)

In [15]:
model = KNeighborsClassifier(n_neighbors=5)
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)

/opt/anaconda3/envs/llm_env/lib/python3.10/site-packages/sklearn/neighbors/_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [16]:
#with open('../data/models/version9.1_KNN(0.62).pkl', 'wb') as file:
#    pickle.dump(model, file)

In [17]:
# Accuracy & F1 Score
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')
print(f'f1_score: {f1:.2f}')

Accuracy: 0.56
f1_score: 0.64


In [18]:
# Classification Report
class_report = classification_report(y_test, y_pred, target_names=['Down', 'Up'])
print('Classification Report:')
print(class_report)

Classification Report:
              precision    recall  f1-score   support

        Down       0.57      0.34      0.42       321
          Up       0.55      0.77      0.64       341

    accuracy                           0.56       662
   macro avg       0.56      0.55      0.53       662
weighted avg       0.56      0.56      0.54       662



## XGBoost

In [19]:
model = xgb.XGBClassifier(n_estimators=100, random_state=42)
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)

In [20]:
#with open('../data/models/version10.1_xgboost(0.64).pkl', 'wb') as file:
#   pickle.dump(model, file)

In [21]:
# Accuracy & F1 Score
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')
print(f'f1_score: {f1:.2f}')

Accuracy: 0.65
f1_score: 0.68


In [22]:
# Classification Report
class_report = classification_report(y_test, y_pred, target_names=['Down', 'Up'])
print('Classification Report:')
print(class_report)

Classification Report:
              precision    recall  f1-score   support

        Down       0.66      0.59      0.62       321
          Up       0.65      0.72      0.68       341

    accuracy                           0.65       662
   macro avg       0.66      0.65      0.65       662
weighted avg       0.65      0.65      0.65       662



## Neural Networks (MLP)

In [23]:
model = MLPClassifier(hidden_layer_sizes=(100,), max_iter=300, random_state=42)
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)


/opt/anaconda3/envs/llm_env/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1105: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/opt/anaconda3/envs/llm_env/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


In [24]:
# Accuracy & F1 Score
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')
print(f'f1_score: {f1:.2f}')

Accuracy: 0.63
f1_score: 0.71


In [25]:
# Classification Report
class_report = classification_report(y_test, y_pred, target_names=['Down', 'Up'])
print('Classification Report:')
print(class_report)

Classification Report:
              precision    recall  f1-score   support

        Down       0.74      0.36      0.48       321
          Up       0.59      0.88      0.71       341

    accuracy                           0.63       662
   macro avg       0.67      0.62      0.60       662
weighted avg       0.67      0.63      0.60       662



In [26]:
#with open('../data/models/version11_MLP(0.58).pkl', 'wb') as file:
#   pickle.dump(model, file)

## Load the Final Data

In [27]:
X_train_scaled = pd.read_csv('../data/final/X_train_scaled.csv')
X_test_scaled = pd.read_csv('../data/final/X_test_scaled.csv')
y_train = pd.read_csv('../data/final/y_train.csv')
y_test = pd.read_csv('../data/final/y_test.csv')

#### Test Logistic Regression Model for Overfitting

In [28]:
model = LogisticRegression()
model.fit(X_train_scaled, y_train)

train_pred = model.predict(X_train_scaled)
test_pred = model.predict(X_test_scaled)

# Calculate F1 scores
train_score = f1_score(y_train, train_pred)
test_score = f1_score(y_test, test_pred)

# Print the results
print("Training F1 Scores: ", train_score)
print("Testing F1 Scores: ", test_score)

Training F1 Scores:  0.7398005439709883
Testing F1 Scores:  0.7329479768786127


/opt/anaconda3/envs/llm_env/lib/python3.10/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Train and Test data are roughly the same, therefore no overfitting.
- Training F1 Scores:  0.7398005439709883
- Testing F1 Scores:  0.7329479768786127


#### Test XGBoost Model for Overfitting

In [29]:
model = xgb.XGBClassifier(n_estimators=100, random_state=42)
model.fit(X_train_scaled, y_train)

train_pred = model.predict(X_train_scaled)
test_pred = model.predict(X_test_scaled)

# Calculate F1 scores
train_score = f1_score(y_train, train_pred)
test_score = f1_score(y_test, test_pred)

# Print the results
print("Training F1 Scores: ", train_score)
print("Testing F1 Scores: ", test_score)

Training F1 Scores:  1.0
Testing F1 Scores:  0.6815020862308763


Test F1 Score is much lower that test F1 Score suggesting the model is overfitting. Further exploration on this will be done in hyperparameter tuning to see if I can address the model's overfitting.
- Training F1 Scores:  1.0
- Testing F1 Scores:  0.6815020862308763